src/layer/unitary.rs

```diff
@@ -0,0 +1,468 @@
+use num;
+use af;
+use af::{Array, MatProp, Dim4, DType};
+
+use utils;
+use std::sync::{Arc, Mutex};
+use activations;
+use initializations;
+use params::{Params};
+use layer::Layer;
+
+use num::Complex;
+
+
+pub struct Unitary {
+    pub input_size: usize,
+    pub output_size: usize,
+}
+
+fn h_d(param: Array, ar: Array) -> Array {
+    let D = af::cplx2(&af::cos(&param)
+                      , &af::sin(&param)
+                      ,false);
+    af::mul(&D, &ar, true)
+}
+
+fn r_d(param: Array, ar: Array) -> Array {
+    let D = af::cplx2(&af::cos(&param)
+                      , &af::mul(&af::sin(&param), &-1, true)
+                      ,false);
+    af::mul(&D, &ar, true)
+}
+
+fn h_fft(ar: Array) -> Array {
+    let ar_size = ar.dims()[1];
+    af::transpose(&af::fft(&af::transpose(&ar, false), 1.0, ar_size as i64)
+                  , false)
+}
+
+fn r_fft(ar: Array) -> Array {
+    let ar_size = ar.dims()[1];
+    af::transpose(&af::ifft(&af::transpose(&ar, false), 1.0, ar_size as i64)
+                  , false)
+}
+
+fn h_ifft(ar: Array) -> Array {
+    let ar_size = ar.dims()[1];
+    af::transpose(&af::ifft(&af::transpose(&ar, false), 1.0/(ar_size as f64), ar_size as i64)
+                  , false)
+}
+
+fn r_ifft(ar: Array) -> Array {
+    let ar_size = ar.dims()[1];
+    af::transpose(&af::fft(&af::transpose(&ar, false), 1.0/(ar_size as f64), ar_size as i64)
+                  , false)
+}
+
+fn h_pi(param: Array, ar: Array) -> Array {
+    af::lookup(&ar, &param, 1)
+}
+
+fn h_r(param: Array, mut ar: Array) -> Array {
+    let ar_temp = ar.clone();
+    ar = af::matmul(&param
+                    , &ar
+                    , MatProp::NONE
+                    , MatProp::TRANS);
+    ar = af::matmul(&ar
+                    , &af::conjg(&param)
+                    , MatProp::TRANS
+                    , MatProp::NONE);
+    ar = af::mul(&ar, &2, true);
+    ar = af::sub(&ar_temp, &ar, false);
+    ar
+
+}
+
+// Wh = (D3 R2 F-1 D2 Pi R1 F D1) h
```

Comment:
Add a little more commentary here.  I suggest following this style: https://github.com/jramapuram/hal/blob/feature/lstm/src/utils.rs#L398-L402

I have started this for the other layers/code as well. If you use this style the docs are auto-generated for us!